### PR TITLE
added dist tag ls/add/rm

### PR DESCRIPTION
**Pull request by dickeyxxx**

### CODE REVIEW

File: lib/npm.js [hunk 1]
```diff
@@ -69,7 +70,14 @@ function getTarball (name, filename) {
   })
 }
 
+function * getDistTags (name) {
+  let opts = {timeout: config.timeout, json: true}
+  let rsp = yield got(url.resolve(config.uplink.href, '/-/package/' + name.replace(/\//, '%2F') + '/dist-tags'), opts)
```
Comment (ransombriggs) [inline lib/npm.js]: @dickeyxxx could you `encodeURIComponent` just in case name has some other problematic characters?

File: routes/dist_tags.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+const npm = require('../lib/npm')
+const config = require('../config')
+const job = require('../lib/job')
+const _ = require('lodash')
+const parse = require('co-body')
+const middleware = require('../middleware')
+
+let updateDistTags = job(co.wrap(function * (name, tags) {
+  let key = `dist-tags/${name}`
+  let current = yield config.storage.getJSON(key)
```
Comment (ransombriggs) [inline routes/dist_tags.js]: @dickeyxxx you should probably use `getStorageTags` here, although is it really necessary to do a check here for equality.  It makes sense if the `put` costs a lot more than the `get` but do not know if that is the case.

File: routes/dist_tags.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  tags[tag] = version
+  yield config.storage.put(`dist-tags/${name}`, tags)
+  this.body = {}
+})
+
+r.delete('/-/package/:name/dist-tags/:tag', middleware.auth, function * () {
+  let {name, tag} = this.params
+  let tags = yield npm.getDistTags(name).catch(() => 'not found')
+  if (tags !== 'not found') this.throw(400, `Cannot set dist-tags, ${name} is hosted on ${config.uplink.host}`)
+  tags = yield getStorageTags(name)
```
Comment (ransombriggs) [inline routes/dist_tags.js]: @dickeyxxx previous four lines have identical lines in the `put`, should probably extract to a function

File: routes/publish.js [hunk 4]
```diff
@@ -11,9 +11,10 @@ const config = require('../config')
 // npm publish
 r.put('/:name', middleware.auth, function * () {
   let pkg = yield parse(this, {limit: '100mb'})
+  let tag = Object.keys(pkg['dist-tags'])[0]
```
Comment (ransombriggs) [inline routes/publish.js]: @dickeyxxx I am assuming that people pass multiple tags at once, should either throw an error here or handle if they pass multiples

File: routes/dist_tags.js (hunk 2 above)
Comment (dickeyxxx) [inline routes/dist_tags.js]: yeah, S3 PUT is much more expensive than GET.

Comment (dickeyxxx) [inline routes/dist_tags.js]: And this is called every time the CLI updates, so it's good to be mindful of it.

File: routes/dist_tags.js (hunk 3 above)
Comment (dickeyxxx) [inline routes/dist_tags.js]: I don't think I agree. The only thing I can see that could be shared between the 2 really is

```
let tags = yield npm.getDistTags(name).catch(() => 'not found')
```

But that's already extracted aside from the catch handler. The rest is grabbing the params out of the request and showing an error message. I think those should stay inside the route.

File: routes/publish.js (hunk 4 above)
Comment (dickeyxxx) [inline routes/publish.js]: there isn't a way to specify more than 1 tag with the CLI

**Comment by dickeyxxx** [comment]
@ransombriggs essentially I need to make all the `npm dist-tag *` commands work with elephant

File: lib/npm.js (hunk 1 above)
Comment (dickeyxxx) [inline lib/npm.js]: done

File: routes/dist_tags.js (hunk 2 above)
Comment (dickeyxxx) [inline routes/dist_tags.js]: updated

**Comment by ransombriggs** [comment]
@dickeyxxx lgtm

**Pull request merged**